parallel.go

```diff
@@ -298,3 +304,137 @@ func (w *parallelManager) Call(data interface{}) error {
 func (w *parallelManager) Wait() (interface{}, error) {
 	return nil, ErrWouldDeadlock
 }
+
+// goWorker is an implementation of the Worker interface that operates
+// in a parallel fashion, utilizing a semaphore to control the number
+// of goroutines allowed to operate at once.
+type goWorker struct {
+	sync.Mutex
+	state  pState              // State of the worker
+	serial *sync.Mutex         // A mutex for serializing Runner.Integrate
+	runner Runner              // The runner to be invoked by the workers
+	gonner *sync.Once          // A once incarnation for getting the result
+	result interface{}         // The result from the work
+	limit  *semaphore.Weighted // Semaphore to limit concurrent execution
+	wg     *sync.WaitGroup     // Wait group to use for waits
+}
+
+// NewGoWorker constructs a worker utilizing a semaphore to limit
+// concurrency of worker goroutines.  Go workers can receive Call and
+// Wait invocations from any goroutine.  A go worker is initialived
+// with a desired maximum number of simultaneously executing
+// goroutines; if that number is less than or equal to 0, no limit is
+// enforced on the number of simultaneous goroutines.
+func NewGoWorker(runner Runner, workers int) Worker {
+	// Initialize a semaphore
+	var sem *semaphore.Weighted
+	if workers > 0 {
+		sem = semaphore.NewWeighted(int64(workers))
+	}
+
+	return &goWorker{
+		serial: &sync.Mutex{},
+		runner: runner,
+		gonner: &sync.Once{},
+		limit:  sem,
+		wg:     &sync.WaitGroup{},
+	}
+}
+
+// work is a helper that executes in a fresh goroutine.  It acquires
+// the semaphore, executes the runner's Run method with the desired
+// data, runs the runner's Integrate method with the result, then
+// dones the wait group.
+func (w *goWorker) work(data interface{}) {
+	// Signal done when we're done
+	defer w.wg.Done()
+
+	// First, acquire the semaphore; this limits the parallelism
+	if w.limit != nil {
+		w.limit.Acquire(context.Background(), 1) // cannot error with background context
+	}
+
+	// Now we can run the runner
+	result := panicer(w.runner.Run, data)
+
+	// Release the semaphore
+	if w.limit != nil {
+		w.limit.Release(1)
+	}
+
+	// Next, lock the serialization mutex
+	w.serial.Lock()
+	defer w.serial.Unlock()
+
+	// Integrate the result
+	w.runner.Integrate(w, result)
+}
+
+// getResult is a helper for Wait to retrieve the result.  It's called
+// with goWorker.gonner to ensure that it only gets called once.
+func (w *goWorker) getResult() {
+	w.Lock()
+	defer w.Unlock()
+
+	w.result = w.runner.Result()
+	w.state = pResult
+}
+
+// Call is the method used to submit data to be worked in a call to
+// the Runner.Run method.  It may return an error if the worker has
+// been shut down through a call to Wait.
+func (w *goWorker) Call(data interface{}) error {
+	// Check the state
+	w.Lock()
+	switch w.state {
+	case pNew: // Need to start up
+		w.state = pRunning
+
+	case pClosed, pResult: // Oh, we're closed
+		return ErrClosed
+	}
```

Comment:
Should there maybe be an explicit do-nothing block for the case where current state is pRunning, or a comment noting that the pRunning state is expected?